File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> prefix = new HashMap<>();
        int total = 0, result = 0;
        prefix.put(0, -1);

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                total--;
            } else {
                total++;
            }
            if (!prefix.containsKey(total)) {
                prefix.put(total, i);
            } else {
                result = Math.max(result, i - prefix.get(total));
            }
        }

        return result;
    }
}

File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        prefix = {0: -1}
        total, result = 0, 0

        for i in range(len(nums)):
            if nums[i] == 0:
                total -= 1
            else:
                total += 1
            if total not in prefix:
                prefix[total] = i
            else:
                result = max(result, i - prefix[total])

        return result